### PR TITLE
feat: add upload document to S3 and list document page

### PR DESCRIPTION
**Pull request by jasonirvine76**
### Test Additions:
- Added a test for the List Document View to ensure the view correctly renders the document list page with the appropriate context data.
- Added a test for the Upload Document Service to ensure documents are uploaded successfully and handle errors appropriately.

### Features Implemented:
- Implemented a feature to display a list of documents, including fetching the required data sources and passing them to the front-end template.
- Upload Document Service: Implemented a service for handling document uploads, with proper error handling for cases such as missing credentials and botocore errors.
- Route for List Document View: A route was added to handle document listing requests.

### CODE REVIEW

**Comment by TakeMeGH** [comment]
I think some monkeypatch for the secret needs to be added to test_config and test_config_default_port to resolve the error.

**Comment by jasonirvine76** [comment]
> I think some monkeypatch for the secret needs to be added to test_config and test_config_default_port to resolve the error.

Thanks for the advice

**Review by nayyara-airlangga — DISMISSED** [review]
This seems like you are building upon Veivel's PR on #58. This is cool but I suggest you change the target branch to Veivel's branch directly. So the overall flow of approval would be something like

1. Approve Veivel's PR and merge
2. The base branch of your PR will automatically change to main afterwards and the commit history will be synced accordingly

**Comment by jasonirvine76** [comment]
> This seems like you are building upon Veivel's PR on #58. This is cool but I suggest you change the target branch to Veivel's branch directly. So the overall flow of approval would be something like
> 
> 1. Approve Veivel's PR and merge
> 2. The base branch of your PR will automatically change to main afterwards and the commit history will be synced accordingly

Alright, will wait for Veivel's PR

**Comment by nayyara-airlangga** [comment]
> > This seems like you are building upon Veivel's PR on #58. This is cool but I suggest you change the target branch to Veivel's branch directly. So the overall flow of approval would be something like
> 
> > 
> 
> > 1. Approve Veivel's PR and merge
> 
> > 2. The base branch of your PR will automatically change to main afterwards and the commit history will be synced accordingly
> 
> 
> 
> Alright, will wait for Veivel's PR 

@jasonirvine76 Oh you don't have to wait until the Veivel's PR is merged. All you have to do is change the base branch of your PR to Veivel's branch

**Comment by nyoosteven** [comment]
Great Job. The documents will be stored in S3 storage, and a new function will be implemented to automate the generation and indexing of the entire document set.

**Comment by Veivel** [comment]
> > > This seems like you are building upon Veivel's PR on #58. This is cool but I suggest you change the target branch to Veivel's branch directly. So the overall flow of approval would be something like
> > 
> > 
> > > 
> > 
> > 
> > > 1. Approve Veivel's PR and merge
> > 
> > 
> > > 2. The base branch of your PR will automatically change to main afterwards and the commit history will be synced accordingly
> > 
> > 
> > Alright, will wait for Veivel's PR
> 
> @jasonirvine76 Oh you don't have to wait until the Veivel's PR is merged. All you have to do is change the base branch of your PR to Veivel's branch

Just to clarify @nayyara-airlangga , do I have to merge first and then @jasonirvine76 ? Or just one of us – I'm assuming if both of us merge to main, there will be conflict because we're commiting my own changes twice?

File: config/__init__.py
Comment (Veivel) [inline config/__init__.py:95]: @jasonirvine76 don't forget to add these entries to `.env.example`

File: document/service.py
Comment (Veivel) [inline document/service.py:52]: I know I showed an example prior, but do you think this might be worth putting to a `.env` config?

**Review by Veivel — CHANGES_REQUESTED** [review]

**Comment by nayyara-airlangga** [comment]
> > > > This seems like you are building upon Veivel's PR on #58. This is cool but I suggest you change the target branch to Veivel's branch directly. So the overall flow of approval would be something like
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > 1. Approve Veivel's PR and merge
> 
> > > 
> 
> > > 
> 
> > > > 2. The base branch of your PR will automatically change to main afterwards and the commit history will be synced accordingly
> 
> > > 
> 
> > > 
> 
> > > Alright, will wait for Veivel's PR
> 
> > 
> 
> > @jasonirvine76 Oh you don't have to wait until the Veivel's PR is merged. All you have to do is change the base branch of your PR to Veivel's branch
> 
> 
> 
> Just to clarify @nayyara-airlangga , do I have to merge first and then @jasonirvine76 ? Or just one of us – I'm assuming if both of us merge to main, there will be conflict because we're commiting my own changes twice?

@Veivel either is fine. But jason's target branch for this PR should be pointing towards yours instead of `main`. Personally, I think merging yours first is better as it simplifies the process

File: document/service.py
Comment (jasonirvine76) [reply document/service.py:52]: Yes thanks for the suggestion

File: config/__init__.py
Comment (jasonirvine76) [reply config/__init__.py:95]: Alright

File: document/service.py
Comment (Veivel) [inline document/service.py:56]: any particular reason for erasing this line? I think it's helpful for debugging

File: main.py
Comment (Veivel) [inline main.py:84]: Minor nitpick: since there's 4 values for AWS-related configs, have you considered turning this into a TypedDict or class (like with GoogleCredentials)?

**Review by Veivel — APPROVED** [review]
Made some comments, but they are nitpicks and not really issues; feel free to make necessary adjustments if you feel they're important. Everything else looks good!

File: document/service.py
Comment (jasonirvine76) [reply document/service.py:56]: Ah yes, i accidentally remove it

File: main.py
Comment (jasonirvine76) [reply main.py]: Done, thanks for the suggestion

**Pull request merged**
